common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDBIterator.java

```diff
@@ -272,4 +269,39 @@ static int compare(byte[] a, byte[] b) {
     return a.length - b.length;
   }
 
+  static class ResourceCleaner implements Runnable {
+
+    private final RocksIterator rocksIterator;
+
+    private final AtomicReference<org.rocksdb.RocksDB> _db;
+
+    private final ConcurrentLinkedQueue<Reference<RocksDBIterator<?>>> iteratorTracker;
+
+    public ResourceCleaner(RocksIterator rocksIterator,
+        AtomicReference<org.rocksdb.RocksDB> _db,
+        ConcurrentLinkedQueue<Reference<RocksDBIterator<?>>> iteratorTracker) {
+      this.rocksIterator = rocksIterator;
+      this._db = _db;
+      this.iteratorTracker = iteratorTracker;
+    }
+
+    @Override
+    public void run() {
+      synchronized (this._db) {
+        org.rocksdb.RocksDB _db = this._db.get();
+        if (_db == null) {
+          return;
+        }
+        rocksIterator.close();
+      }
+      iteratorTracker.removeIf(ref -> {
+          RocksDBIterator<?> rocksDBIterator = ref.get();
+          if (rocksDBIterator != null && rocksIterator.equals(rocksDBIterator.it)) {
+            return true;
+          } else {
+            return false;
+          }
+        });
```

Comment:
It seems we should remove reference for iteratorTracker first, then close rocksIterator?

Reply:
+1

Reply:
thanks, fix